### PR TITLE
Support sanitising the URL by removing extra slashes in the URL

### LINKED ISSUE #20462: Extra slash in URL makes git clone fail
### Description

Cloning a repository with a double slash fails:
```shell
$ git clone https://git.data.coop//halfd/new-website.git
Cloning into 'new-website'...
remote: Not found.
fatal: repository 'https://git.data.coop//halfd/new-website.git/' not found
```

### Gitea Version

1.16.8

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

Using the docker image.

### Database

SQLite

### LINKED ISSUE #23242: Encoded path will be decoded by `middleware.StripSlashes` and mismatch routers
### Description

When everything is right:

- `https://gitea.com/gitea/blog` -> 200
- `https://gitea.com/gitea%2Fblog` -> 404

However:

- `https://gitea.com/gitea%2Fblog/` -> 200

It is caused by a bug in `middleware.StripSlashes` of chi, see https://github.com/go-chi/chi/pull/799

We don't have to wait for it to be fixed in upstream, it can be fixed in #21333 by dropping `middleware.StripSlashes`.

### Gitea Version

7a5af2559

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

command

### Database

None

### PR DESCRIPTION
**Pull request by sandyydk**
Changes in this PR :

Strips incoming request URL of additional slashes (/). For example an input like 

`https://git.data.coop//halfd/new-website.git` is translated to `https://git.data.coop/halfd/new-website.git`

Fixes https://github.com/go-gitea/gitea/issues/20462

Fix #23242

### CODE REVIEW

**Comment by sandyydk** [comment]
@Gusted Please take a look at this PR and let me know. Thanks. Looking forward to my first commit :)

**Comment by zeripath** [comment]
Unfortunately you've linked in the version 1 chi api - which won't work with chi v5. You'll need to use v5 libraries.

**Comment by sandyydk** [comment]
> Unfortunately you've linked in the version 1 chi api - which won't work with chi v5. You'll need to use v5 libraries.

Thanks. Was trying to see why the test cases were failing for some scenarios. This makes sense.

File: routers/common/middleware.go [hunk 1]
```diff
@@ -81,3 +83,31 @@ func Middlewares() []func(http.Handler) http.Handler {
...
+		rctx := chi.RouteContext(req.Context())
+		if rctx != nil && rctx.RoutePath != "" {
+			urlPath = rctx.RoutePath
+		} else {
+			urlPath = req.URL.Path
+		}
+
+		var sanitizedPath string
+		if len(urlPath) > 1 {
+			sanitizedPath = path.Clean(urlPath)
```
Comment (Gusted) [inline routers/common/middleware.go]: [`path.Clean`](https://pkg.go.dev/path#Clean) is the wrong function for this job. It has other side-effects involving with how file systems works, which Gitea "shouldn't" care about. 
This increased restrictiveness is causing the tests to fails.

File: routers/common/middleware.go [hunk 2]
```diff
@@ -81,3 +83,31 @@ func Middlewares() []func(http.Handler) http.Handler {
...
+
+			if string(urlPath[len(urlPath)-1]) == "/" {
+				sanitizedPath += "/"
+			}
```
Comment (Gusted) [inline routers/common/middleware.go]: The previous code explicitly was removing a trailing slash. Why is this being added back in?

**Review by Gusted — CHANGES_REQUESTED** [review]
Please have a look at [the patches I had made](https://github.com/go-gitea/gitea/issues/20462#issuecomment-1193196481) (under "Details" text) to see how the code likely should look like. The current code is too restrictive is adding back in a trailing slash.

File: routers/common/middleware.go (hunk 1 above)
Comment (sandyydk) [reply routers/common/middleware.go]: @Gusted Sure yeah. Removed it's usage.

File: routers/common/middleware.go (hunk 2 above)
Comment (sandyydk) [reply routers/common/middleware.go]: I added this back in thinking this was causing the test case to fail. Just to test it out once.

File: routers/common/middleware.go (hunk 1 above)
Comment (sandyydk) [reply routers/common/middleware.go]: @Gusted The tests still fail.

Comment (Gusted) [reply routers/common/middleware.go]: I see, never ran tests on non-ascii characters. Code will need adjustments for that likely, you can add some debug in here and run the test via: `make test-sqlite#TestNonasciiBranches` to see what's going on.

**Comment by wolfogre** [comment]
Hi @sandyydk, thank you for your work on this. I noticed that this PR hasn't been updated in a while, and I believe it may be because the maintainers haven't responded for some time. We apologize for any oversight on our part.

Could you please make some updates to resolve the CI failures? I will also call on others to review this PR. If you are too busy to do this or to reply, I can help to modify your code after one day.

**Comment by sandyydk** [comment]
Hi Jason

Sure. Let me check it out and get back to you. Thanks for responding.

File: routers/common/middleware.go [hunk 3]
```diff
@@ -81,3 +82,44 @@ func Middlewares() []func(http.Handler) http.Handler {
...
+
+			// Always remove the leading slash.
+			modifiedPath := strings.TrimSuffix(string(sanitizedPath), "/")
+
+			if rctx == nil {
+				req.URL.Path = modifiedPath
+			} else {
+				rctx.RoutePath = modifiedPath
+			}
+		}
```
Comment (wolfogre) [inline routers/common/middleware.go]:
```suggestion
		sanitizedPath := &strings.Builder{}
		prevWasSlash := false
		for _, chr := range strings.TrimRight(urlPath, "/") {
			if chr != '/' || !prevWasSlash {
				sanitizedPath.WriteRune(chr)
			}
			prevWasSlash = chr == '/'
		}

		if rctx == nil {
			req.URL.Path = sanitizedPath.String()
		} else {
			rctx.RoutePath = sanitizedPath.String()
		}
```

File: routers/common/middleware_test.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
```
Comment (wolfogre) [inline routers/common/middleware_test.go]:
```suggestion
// SPDX-License-Identifier: MIT
```

File: routers/common/middleware.go [hunk 5]
```diff
@@ -81,3 +82,31 @@ func Middlewares() []func(http.Handler) http.Handler {
...
+
+func stripSlashesMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(resp http.ResponseWriter, req *http.Request) {
+		var urlPath string
+		rctx := chi.RouteContext(req.Context())
+		if rctx != nil && rctx.RoutePath != "" {
+			urlPath = rctx.RoutePath
+		} else {
+			urlPath = req.URL.Path
+		}
```
Comment (wolfogre) [inline routers/common/middleware.go]:
```suggestion
		if rctx != nil && rctx.RoutePath != "" {
			urlPath = rctx.RoutePath
		} else if req.URL.RawPath != "" {
			urlPath = req.URL.RawPath
		} else {
			urlPath = req.URL.Path
		}
```

**Review by wolfogre — APPROVED** [review]

File: routers/common/middleware_test.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+		{
+			name:         "path with slashes and query params",
+			inputPath:    "/repo//migrate?service_type=3",
+			expectedPath: "/repo/migrate?service_type=3",
+		},
+	}
+
+	for _, tt := range tests {
+		testMiddleware := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			assert.Equal(t, tt.expectedPath, r.URL.String())
```
Comment (wolfogre) [inline routers/common/middleware_test.go]:
```suggestion
	tests := []test{
		{
			name:         "path with multiple slashes",
			inputPath:    "https://github.com///go-gitea//gitea.git",
			expectedPath: "/go-gitea/gitea.git",
		},
		{
			name:         "path with no slashes",
			inputPath:    "https://github.com/go-gitea/gitea.git",
			expectedPath: "/go-gitea/gitea.git",
		},
		{
			name:         "path with slashes in the middle",
			inputPath:    "https://git.data.coop//halfd/new-website.git",
			expectedPath: "/halfd/new-website.git",
		},
		{
			name:         "path with slashes in the middle",
			inputPath:    "https://git.data.coop//halfd/new-website.git",
			expectedPath: "/halfd/new-website.git",
		},
		{
			name:         "path with slashes in the end",
			inputPath:    "/user2//repo1/",
			expectedPath: "/user2/repo1",
		},
		{
			name:         "path with slashes and query params",
			inputPath:    "/repo//migrate?service_type=3",
			expectedPath: "/repo/migrate",
		},
		{
			name:         "path with encoded slash",
			inputPath:    "/user2/%2F%2Frepo1",
			expectedPath: "/user2/%2F%2Frepo1",
		},
	}

	for _, tt := range tests {
		testMiddleware := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			assert.Equal(t, tt.expectedPath, r.URL.Path)
```

**Review by lunny — APPROVED** [review]

**Pull request merged**